### PR TITLE
Cleanup and extract Transaction Workflow logic

### PR DESCRIPTION
**Pull request by briandamaged**
This extracts the "operational" logic related to container / non-container Transactions.  Saying it another way: it means that the `Transaction` class no longer needs to be aware of "how" it was instantiated.

### CODE REVIEW

File: lib/transaction.js [hunk 1]
```diff
@@ -14,35 +14,12 @@ const { uniqueId, isUndefined } = require('lodash');
 // Acts as a facade for a Promise, keeping the internal state
 // and managing any child transactions.
 class Transaction extends EventEmitter {
-  constructor(client, container, config, outerTx) {
+  constructor(client, container, config = {}, outerTx = null) {
     super();
+    this.doNotRejectOnRollback = config.doNotRejectOnRollback;
```
Comment (kibertoad) [inline lib/transaction.js]: Doesn't this change default value to `false` from current `true`?

File: lib/transaction.js [hunk 2]
```diff
@@ -225,12 +203,11 @@ class Transaction extends EventEmitter {
...
 
-        return (this._previousSibling
-          ? this._previousSibling.catch(() => {})
-          : Promise.resolve()
-        ).then(function() {
-          return connection;
-        });
+        return this._previousSibling
+          // .catch(() => {}) // TODO: Investigate this line
+          .then(function() {
```
Comment (kibertoad) [inline lib/transaction.js]: probably can be replaced with async/await?

File: lib/transaction.js (hunk 1 above)
Comment (briandamaged) [reply lib/transaction.js]: Yes and No.  It appears that `Transaction` construction is always kicked off from `lib/util/make-knex.js` .  The logic for handling defaults has been extracted to that location.

But... to be on the safe side, we might also want to provide an ULTRA-default value within the `Transaction` constructor as well.

Comment (kibertoad) [reply lib/transaction.js]: Right, I see, you handle default values elsewhere now.

File: lib/transaction.js (hunk 2 above)
Comment (briandamaged) [reply lib/transaction.js]: It's worth a shot, at least.  My main concern was that `Transaction` itself is a Promise-like object.  So, we need to make sure it isn't partially-resolved by accident.

File: lib/util/make-knex.js [hunk 3]
```diff
@@ -36,17 +36,32 @@ function initContext(knexFn) {
...
+      // Backwards-compatibility: the default value of `config.doNotRejectOnRollback`
+      // changes depending upon whether or not a `container` was provided.
+      if(isUndefined(config.doNotRejectOnRollback)) {
+        config.doNotRejectOnRollback = !container;
+      }
 
-      if (container) {
+      if(container) {
+        const trx = this.client.transaction(container, config);
+        trx.userParams = this.userParams;
```
Comment (briandamaged) [inline lib/util/make-knex.js]: This seems a bit strange to me.  Is there a reason that `userParams` cannot be passed in via the `config` object?

File: lib/transaction.js (hunk 2 above)
Comment (briandamaged) [reply lib/transaction.js]: Actually: Just realized that this PR is showing items that were merged from the other PR (#3671 ) .  I might need to rebase the branch and create a new PR instead.

File: lib/util/make-knex.js (hunk 3 above)
Comment (kibertoad) [reply lib/util/make-knex.js]: They probably could, `userParams` is just a slightly special beast that we use for handling e. g. enabling/disabling preprocessing of queries in addition to its primary purpose of parametrizing specific copy of knex builder :)

**Comment by briandamaged** [comment]
Closing this PR out because there were duplicate commits in its history.

The rebased version of this PR can be found here:  #3674

**Pull request closed without merging**